### PR TITLE
JAMES-2096 Webadmin triggered migrations

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+        service.post(VERSION_UPGRADE_TO_LATEST_BASE, (request, response) -> {
+            try {
+                cassandraMigrationService.upgradeToLastVersion();
+            } catch (IllegalStateException e) {
+                LOGGER.info("The migration requested can not be performed.", e);
+                response.status(MIGRATION_CAN_NOT_BE_PERFORMED);
+                response.body(e.getMessage());
+            } catch (MigrationException e) {
+                LOGGER.info("An error lead to partial migration process", e);
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/routes/CassandraMigrationRoutes.java]: idem?

File: server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/service/CassandraMigrationService.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            LOG.info("Migrating to version {} ", version + 1);
+            boolean migrationSuccess = allMigrationClazz.get(version).run();
+            if (migrationSuccess) {
+                schemaVersionDAO.updateVersion(version + 1);
+                LOG.info("Migrating to version {} done", version + 1);
+            } else {
+                String message = String.format("Migrating to version %d partially done. " +
+                    "Please check logs for cause of failure and re-run this migration.",
+                    version + 1);
+                LOG.info(message);
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-cassandra/src/main/java/org/apache/james/webadmin/service/CassandraMigrationService.java]: info only? maybe warn?

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        when(schemaVersionDAO.getCurrentSchemaVersion()).thenReturn(CompletableFuture.completedFuture(Optional.of(LATEST_VERSION)));
+
+        when()
+            .post("/upgrade/latest")
+        .then()
+            .statusCode(410);
+
+        verify(schemaVersionDAO, times(1)).getCurrentSchemaVersion();
+        verifyNoMoreInteractions(schemaVersionDAO);
+    }
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: do you handle the case where no current version is stored in Cassandra?

File: server/container/guice/protocols/webadmin-cassandra/src/main/java/org/apache/james/modules/server/CassandraRoutesModule.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+import com.google.inject.AbstractModule;
+import com.google.inject.Scopes;
+import com.google.inject.multibindings.MapBinder;
+import com.google.inject.multibindings.Multibinder;
+import com.google.inject.name.Names;
+
+public class CassandraRoutesModule extends AbstractModule {
+
+    private static final int FROM_V1_TO_V2 = 1;
```
Comment (rouazana) [inline server/container/guice/protocols/webadmin-cassandra/src/main/java/org/apache/james/modules/server/CassandraRoutesModule.java]: I don't understand why FROM_V1_TO_V2 is equal to 1.

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2MigrationTest.java [hunk 5]
```diff
@@ -166,6 +166,52 @@ public void migrationShouldWorkWithoutAttachments() throws Exception {
...
+
+        CassandraMessageDAOV2.MessageResult messageResult = retrieveMessageOnV2().get();
+        softly.assertThat(messageResult.message().getLeft().getMessageId()).isEqualTo(messageId);
+        softly.assertThat(IOUtils.toString(messageResult.message().getLeft().getContent(), Charsets.UTF_8))
+            .isEqualTo(CONTENT);
+        softly.assertThat(messageResult.message().getRight().findAny().isPresent()).isFalse();
+    }
+
+    @Test
+    public void ruwShouldSucceedWhenOneMessage() throws Exception {
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2MigrationTest.java]: more ruw :)

File: mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2MigrationTest.java [hunk 6]
```diff
@@ -166,6 +166,52 @@ public void migrationShouldWorkWithoutAttachments() throws Exception {
...
+        SimpleMailboxMessage originalMessage = createMessage(messageId, CONTENT, BODY_START,
+            new PropertyBuilder(), ImmutableList.of());
+        messageDAOV1.save(originalMessage).join();
+
+        assertThat(testee.run()).isTrue();
+    }
+
+    @Test
+    public void ruwShouldSucceedWhenNoMessages() throws Exception {
```
Comment (rouazana) [inline mailbox/cassandra/src/test/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2MigrationTest.java]: and here

File: src/site/markdown/server/manage-webadmin.md [hunk 7]
```diff
@@ -331,4 +331,93 @@ Response codes:
...
  - 400: Invalid JSON, or numbers are less than -1.
- - 500: Internal error
\ No newline at end of file
+ - 500: Internal error
+
+## Cassandra Schema upgrades
+
+Cassandra upgrades implies the creation of a new table. Thus restarting James is needed, as new tables are created on restart.
+
+Once done, we ship code that tries to read from new tables, and if not possible backs up to old tables. You can can thus safely run
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: can can

File: src/site/markdown/server/manage-webadmin.md [hunk 8]
```diff
@@ -331,4 +331,93 @@ Response codes:
...
+without running additional migrations.
+
+On the fly migration can be enabled. However, one might want to force the migration in a controlled fashion, and update
+automatically current schema version use (assess in the database old versions is no more used, as the corresponding tables are empty).
+Note the this process is safe: we ensure the service is not running concurrently on this James instance, that it do not bump
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: s/Note the/Note that
s/it do not/it does not

File: src/site/markdown/server/manage-webadmin.md [hunk 9]
```diff
@@ -331,4 +331,93 @@ Response codes:
...
+curl -XPOST http://ip:port/cassandra/version/upgrade -d '3'
+```
+
+Will run the migrations you need to reach schema version 3.
+
+Response codes:
+
+ - 204: Success
+ - 400: The version is invalid. The version should be a strictly positive number.
+ - 410: Error running the migration. This resource is gone away. Reason is metionned in the body.
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: s/metionned/mentioned

File: src/site/markdown/server/manage-webadmin.md [hunk 10]
```diff
@@ -331,4 +331,93 @@ Response codes:
...
+
+Will run the migrations you need to reach the latest schema version.
+
+Response codes:
+
+ - 204: Success
+ - 410: Error running the migration. This resource is gone away. Reason is metionned in the body.
+ - 500: Internal error. This can result from a partial migration, in which case the reason is contained in the body.
+
+Note that several calls to this endpoint will be run in a sequential pattern.
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: can you explain what happends if the server is restarted during the migration?

File: src/site/xdoc/server/config-cassandra.xml [hunk 11]
```diff
@@ -126,6 +126,22 @@
 
   </section>
 
+  <section name="Cassandra migration process">
+    <p>Cassandra upgrades implies the creation of a new table. Thus restarting James is needed, as new tables are created on restart.</p>
+
+    <p>Once done, we ship code that tries to read from new tables, and if not possible backs up to old tables. You can can thus safely run
```
Comment (rouazana) [inline src/site/xdoc/server/config-cassandra.xml]: can can

File: src/site/xdoc/server/config-cassandra.xml [hunk 12]
```diff
@@ -126,6 +126,22 @@
...
+    without running additional migrations.</p>
+
+    <p>On the fly migration can be enabled. However, one might want to force the migration in a controlled fashion, and update
+    automatically current schema version use (assess in the database old versions is no more used, as the corresponding tables are empty).
+    Note the this process is safe: we ensure the service is not running concurrently on this James instance, that it do not bump
```
Comment (rouazana) [inline src/site/xdoc/server/config-cassandra.xml]: Note the/that

Comment (rouazana) [inline src/site/xdoc/server/config-cassandra.xml]: do/does

File: src/site/xdoc/server/config-spring-jpa-postgres.xml [hunk 13]
```diff
@@ -42,7 +42,7 @@
 
         <p>Example to install it on Debian:</p>
         <source>
-          sudo apt-get update && sudo apt-get install postgresql-9.4 postgresql-client-9.4
+          sudo apt-get update &amp;&amp; sudo apt-get install postgresql-9.4 postgresql-client-9.4
```
Comment (rouazana) [inline src/site/xdoc/server/config-spring-jpa-postgres.xml]: unrelated commit?

File: src/site/markdown/server/manage-webadmin.md [hunk 14]
```diff
@@ -331,4 +331,93 @@ Response codes:
...
+
```
Comment (rouazana) [inline src/site/markdown/server/manage-webadmin.md]: could have been a good candidate for swagger :trollface:

File: mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2Migration.java [hunk 15]
```diff
@@ -104,4 +104,9 @@ public void stop() {
         }
         return messageV1;
     }
+
+    @Override
+    public void run() {
```
Comment (chibenwa) [inline mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/migration/V1ToV2Migration.java]: boolean.

Sorry, rebase mistake

File: server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java (hunk 3 above)
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-cassandra/src/test/java/org/apache/james/webadmin/routes/CassandraMigrationRoutesTest.java]: A current version is always returned, and is defined in CassandraSchemaVersionManager

File: server/container/guice/protocols/webadmin-cassandra/src/main/java/org/apache/james/modules/server/CassandraRoutesModule.java (hunk 4 above)
Comment (chibenwa) [inline server/container/guice/protocols/webadmin-cassandra/src/main/java/org/apache/james/modules/server/CassandraRoutesModule.java]: The index quynh chose is the one of the origin version.

File: src/site/xdoc/server/config-spring-jpa-postgres.xml (hunk 13 above)
Comment (chibenwa) [inline src/site/xdoc/server/config-spring-jpa-postgres.xml]: It avoided me generating the site...

File: src/site/markdown/server/manage-webadmin.md (hunk 14 above)
Comment (chibenwa) [inline src/site/markdown/server/manage-webadmin.md]: I like uncoupling issues !

**Comment by chibenwa** [comment]
Rebase was needed due to bad commit message, and a previous rebase mistake.

See the rebased version here: https://github.com/linagora/james-project/pull/907
